core/src/replay_stage.rs

```diff
@@ -783,37 +799,93 @@ impl ReplayStage {
         ancestors: &mut HashMap<Slot, HashSet<Slot>>,
         descendants: &mut HashMap<Slot, HashSet<Slot>>,
         progress: &mut ProgressMap,
         bank_forks: &RwLock<BankForks>,
+        blockstore: &Blockstore,
+        poh_bank_slot: Option<Slot>,
     ) {
-        for duplicate_slot in duplicate_slots_reset_receiver.try_iter() {
-            Self::purge_unconfirmed_duplicate_slot(
-                duplicate_slot,
-                ancestors,
-                descendants,
-                progress,
-                bank_forks,
-            );
+        if duplicate_slots_to_repair.is_empty() {
+            return;
         }
+
+        let root_bank = bank_forks.read().unwrap().root_bank();
+        // TODO: handle if alternate version of descendant also got confirmed after ancestor was
+        // confirmed, what happens then? Should probably keep track of purged list and skip things
+        // in `duplicate_slots_to_repair` that have already been purged. Add test.
+        duplicate_slots_to_repair.retain(|(duplicate_slot, _correct_hash)| {
+            // Should not purge duplicate slots if there is currently a poh bank building
+            // on top of that slot, as BankingStage might still be referencing/touching that state
+            // concurrently.
+            // Luckily for us, because the fork choice rule removes duplicate slots from fork
+            // choice, and this function is called after:
+            // 1) We have picked a bank to reset to in `select_vote_and_reset_forks()`
+            // 2) And also called `reset_poh_recorder()`
+            // Then we should have reset to a fork that doesn't include the duplicate block,
+            // which means any working bank in PohRecorder that was built on that duplicate fork
+            // should have been cleared as well. However, if there is some violation of this guarantee,
+            // then log here
+            let is_poh_building_on_duplicate_fork = poh_bank_slot
+                .map(|poh_bank_slot| {
+                    ancestors
+                        .get(&poh_bank_slot)
+                        .expect("Poh bank should exist in BankForks and thus in ancestors map")
+                        .contains(&duplicate_slot)
+                })
+                .unwrap_or(false);
+
+            let did_purge_repair = {
+                if !is_poh_building_on_duplicate_fork {
+                    Self::purge_unconfirmed_duplicate_slot(
+                        *duplicate_slot,
+                        ancestors,
+                        descendants,
+                        progress,
+                        &root_bank,
+                        bank_forks,
+                        blockstore,
+                    );
+                    true
+                // TODO: Send signal to repair to repair the correct version of
+                // `duplicate_slot` with hash == `correct_hash`
+                } else {
```

Comment:
We don't actually signal repair to go actively repair this duplicate slot, so right now repair will go through the passive path. This means we could potentially repair from the bad set of validators again and get the wrong version again.

Reply:
So we retain the duplicate slots for which the PoH bank is building on, and then later if we reset PoH to a different slot we will go ahead and repair. Does this order of events cause issues?